src/include/sof/lib/ams_msg.h

```diff
@@ -0,0 +1,27 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Copyright (c) 2023 Intel Corporation
+ *
+ * Author: Ievgen Ganakov <ievgen.ganakov@intel.com>
+ */
+
+#ifndef __SOF_LIB_AMS_MSG_H__
+#define __SOF_LIB_AMS_MSG_H__
+
+/* AMS messages */
+
+/* Key-phrase detected AMS message uuid: 80a11122-b36c-11ed-afa1-0242ac120002*/
+#define AMS_KPD_MSG_UUID 0x80a11122, 0xb36c, 0x11ed, 0xaf, 0xa1, 0x02, 0x42, 0xac, 0x12, 0x00, 0x02
```

Comment:
I don't understand what problem DEFINE_UUID() solves. How about something much simpler like this:

```C
typedef uint8_t simple_uuid_t[16];
#define AMS_KPD_MSG_UUID { 0x80, 0xa1, 0x11, 0x22, 0xb3, 0x6c, 0x11, 0xed, 0xaf, 0xa1, 0x02, 0x42, 0xac, 0x12, 0x00, 0x02 }

static const simple_uuid_t ams_kpd_msg_uuid = AMS_KPD_MSG_UUID;
```

Reply:
In case alignement was a concern:
```C
typedef union {
  uint8_t bytes[16];
  uint32_t _; // just for alignment
} simple_uuid_t;

#define AMS_KPD_MSG_UUID { 0x80, 0xa1, 0x11, 0x22, 0xb3, 0x6c, 0x11, 0xed, 0xaf, 0xa1, 0x02, 0x42, 0xac, 0x12, 0x00, 0x02 }

static const simple_uuid_t ams_kpd_msg_uuid = {
  .bytes = AMS_KPD_MSG_UUID
};

```

Reply:
> This gives a checkpatch error "ERROR: Macros with complex values should be enclosed in parentheses" , but I think it's ok. At least I can't come up with a better way to wrap this. @lyakh ideas on this?

@kv2019i I actually think including `static` in `DECLARE_SOF_UUID` isn't a very good idea. I'd remove `static` from it and add it in every specific use like `static DECLARE_SOF_UUID(...)`, and move the `__section()` attribute to a more traditional place, e.g. after the name. Would that allow AMS use the same macro then? And in fact `DEFINE_UUID_` also has `static` in it, which shouldn't be done probably either. So, if at all possible, would be best to use the common SOF UUID tooling. But then the problem remains if we really still need a macro with just the values. I think yes, it's ok to ignore the checkpatch error in this case.

Reply:
I can do it the way @marc-hb suggested. Can we agree on that @lyakh @kv2019i?

Reply:
is it impossible to re-use `DECLARE_SOF_UUID()`? What would be the problem with it - apart from `static`?

Reply:
@lyakh AMS message uuid should be defined in common header. There are at least 4 modules which will use AMS - KPB, WHM, WoV, AEC, more are coming. There will be more then 10 messages defined and 3rd party modules should be able to import them. If I'll use a macro like @lgirdwood suggested I won't be able to path its value to `DECLARE_SOF_UUID()` without splitting it to proper args. So I suppose additional macro would be needed.

Reply:
why not have an extern GUID_BLAH in the header and the value stored in a variable? Macros are evil.

Reply:
> So, if at all possible, would be best to use the common SOF UUID tooling.

The entire idea behind UUIDs is that they can be universally exchanged across all projects/files/modules/languages/etc. without having to share or re-use anything: no header file, no API, no macro, no naming convention, no central registration,... and not tooling. Yet somehow over-engineering manages to creep in anyway :-(

I agree we shouldn't copy/paste ` __section(".static_uuids")` or solutions to endianness issues BUT if some files somewhere only want to share some universal 16 bytes constant (either macro or extern) and nothing else then these files shouldn't have to worry about re-using some "tooling"; that would be overkill.

Reply:
> without having to share or re-use anything: no header file, no API,...

Another key feature of UUIDs is IMMUTABILITY. I hate duplication more than anyone else but immutability means UUID values _can_ be and in fact are copied if/when needed. Not just for fun of course but if/when it makes things easier.

“A little copying is better than a little dependency”

Reply:
> If I'll use a macro like @lgirdwood suggested I won't be able to path its value to `DECLARE_SOF_UUID()` without splitting it to proper args. So I suppose additional macro would be needed.

@iganakov I've just tried this and it seems to have worked. Whether or not we decide to define UUIDs in headers, it seems like it's also possible with the existing API:
```
diff --git a/src/lib/notifier.c b/src/lib/notifier.c
index a601538d7..ccb524ff6 100644
--- a/src/lib/notifier.c
+++ b/src/lib/notifier.c
@@ -22,8 +22,9 @@
 LOG_MODULE_REGISTER(notifier, CONFIG_SOF_LOG_LEVEL);
 
 /* 1fb15a7a-83cd-4c2e-8b32-4da1b2adeeaf */
-DECLARE_SOF_UUID("notifier", notifier_uuid, 0x1fb15a7a, 0x83cd, 0x4c2e,
-		 0x8b, 0x32, 0x4d, 0xa1, 0xb2, 0xad, 0xee, 0xaf);
+#define uuid_notify 0x1fb15a7a, 0x83cd, 0x4c2e, 0x8b, 0x32, 0x4d, 0xa1, 0xb2, 0xad, 0xee, 0xaf
+#define __DECLARE_SOF_UUID(...) DECLARE_SOF_UUID(__VA_ARGS__)
+__DECLARE_SOF_UUID("notifier", notifier_uuid, uuid_notify);
 
 DECLARE_TR_CTX(nt_tr, SOF_UUID(notifier_uuid), LOG_LEVEL_INFO);
 
```

Reply:
@lyakh I've tried your approach and I'm unable to access uuid value in runtime using `DECLARE_SOF_UUID()`. Then I noticed the macro description which tells there is need of uuid dictionary from ldc file to do that. Is there a way to avoid this using SOF uuid infrastructure?

Reply:
Indeed `uuid.h` seems to have grown some coupling with SMEX and .ldc files which the most advanced Zephyr configurations... don't even use anymore.

Again the entire purpose and design of UUIDs is be "universal" and simple by NOT requiring any API/framework/tooling. If the `DECLARE_SOF_UUID()` "tooling" does not work in this particular case then don't use any tooling at all because none is required yet everything will STILL be compatible with anything else using the same UUIDs. That's the magic of UUIDs: compatibility with zero API.